report_py3o/py3o_parser.py

```diff
@@ -0,0 +1,217 @@
+                    tmpl_data = tmpl.read()
+
+        if tmpl_data is None:
+            # if for any reason the template is not found
+            raise TemplateNotFound(
+                _(u'No template found. Aborting.'),
+                sys.exc_info(),
+            )
+
+        return tmpl_data
+
+    def _extend_parser_context(self, parser_instance, report_xml):
+        # add default extenders
+        for fct in _extender_functions.get(None, []):
+            fct(report_xml, parser_instance.localcontext)
+        # add extenders for registered on the template
+        xml_id = report_xml.get_external_id().get(report_xml.id)
+        if xml_id in _extender_functions:
+            for fct in _extender_functions[xml_id]:
+                fct(report_xml, parser_instance.localcontext)
+
+    def create_single_pdf(self, cr, uid, ids, data, report_xml, context=None):
+        """ Overide this function to generate our py3o report
+        """
+        if report_xml.report_type != 'py3o':
+            return super(Py3oParser, self).create_single_pdf(
+                cr, uid, ids, data, report_xml, context=context
+            )
+
+        parser_instance = self.parser(cr, uid, self.name2, context=context)
+        parser_instance.set_context(
+            self.getObjects(cr, uid, ids, context),
+            data, ids, report_xml.report_type
+        )
+        self._extend_parser_context(parser_instance, report_xml)
+
+        tmpl_data = self.get_template(report_xml)
+
+        in_stream = StringIO(tmpl_data)
+        out_stream = StringIO()
+        template = Template(in_stream, out_stream)
+        localcontext = parser_instance.localcontext
+        if report_xml.py3o_is_local_fusion:
+            template.render(localcontext)
+            in_stream = out_stream
+            datadict = {}
+        else:
+            expressions = template.get_all_user_python_expression()
+            py_expression = template.convert_py3o_to_python_ast(expressions)
+            convertor = Py3oConvertor()
+            data_struct = convertor(py_expression)
+            datadict = data_struct.render(localcontext)
+
+        filetype = report_xml.py3o_filetype
+        is_native = Formats().get_format(filetype).native
+        if is_native:
+            res = out_stream.getvalue()
+        else:  # Call py3o.server to render the template in the desired format
+            in_stream.seek(0)
+            files = {
+                'tmpl_file': in_stream,
+            }
+            fields = {
+                "targetformat": filetype,
+                "datadict": json.dumps(datadict),
+                "image_mapping": "{}",
+            }
+            if report_xml.py3o_is_local_fusion:
+                fields['skipfusion'] = '1'
+            r = requests.post(
+                report_xml.py3o_server_id.url, data=fields, files=files)
+            if r.status_code != 200:
+                # server says we have an issue... let's tell that to enduser
```

Comment:
This comment does not add anything to the meaning of the code, the raising of an error implicitly says you are telling the user about an error. No need to repeat yourself IMO